tools/autograd/derivatives.yaml

```diff
@@ -1456,6 +1456,10 @@
   self: grad.transpose(dim0, dim1)
   result: auto_linear
 
+- name: transpose_copy(Tensor self, int dim0, int dim1) -> Tensor
+  self: grad.clone().transpose_copy(dim0, dim1)
```

Comment:
Why `clone()` is needed here if `transpose_copy` does the copy?

Reply:
Without `clone()`, few tests (test_variant_consistency_eage, test_variant_consistency_jit_transpose, test_conj_view, test_neg_view) fail with a message:
```
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation
```

Reply:
Does the error appear if `at::transpose_copy` is used instead of the method?
That way we're doing two copies: one here and another inside the `transpose_copy` call.

Reply:
Yes, using `at::transpose_copy(grad, dim0, dim1)` leads to the same error.
However, when using `grad.clone().transpose(dim0, dim1)`, tests pass.

Reply:
I think something is wrong with the `transpose_copy` then. It would be good to create a code snippet to reproduce the error.
The derivative entry now should be `self: grad.clone().transpose(dim0, dim1)` to avoid multiple copies.

Reply:
Here's a reproducer (when using `self: grad.transpose_copy(dim0, dim1)`):
```python
>>> x = torch.tensor([[1, 2], [3, 4]], dtype=torch.float32, requires_grad=True)
>>> y = torch.transpose_copy(x, 0, 1)
>>> y.sum().backward(retain_graph=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pearu/git/pytorch/pytorch-transpose/torch/_tensor.py", line 352, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/pearu/git/pytorch/pytorch-transpose/torch/autograd/__init__.py", line 173, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: unsupported operation: more than one element of the written-to tensor refers to a single memory location. Please clone() the tensor before performing the operation.
```

Reply:
The runtime error is raised when calling `out.copy_`.
Now using `self: grad.clone().transpose(dim0, dim1)`.

Reply:
Is it possible to trigger the same error in normal code, without backward? Can someone potentially hit it? If yes what would be the fix?

Reply:
Do you mean by "normal code" when just using `transpose_copy` calls? If so, I don't have a definite answer other than test_ops.py covers testing `transpose_copy` in a number of ways.

Digging a little further, the runtime error may be a result of some transpose-state test that gives a false-positive result, after all, the `out` and `self`  in `transcopy_copy` ought to be tensors with non-overlapping storage.

Reading the message  "more than one element of the written-to tensor..", could this be related to autograd machinery in some way? @albanD , do you have any ideas?

Reply:
Not sure where it comes from, but why isn't the formula just: `grad.transpose(dim0, dim1)` ?

Reply:
Not sure where it comes from, but why isn't the formula just: `grad.transpose(dim0, dim1)` ?

Reply:
@albanD , see https://github.com/pytorch/pytorch/pull/68867#discussion_r756395212

Reply:
Ho I though this comment was when using the copy version of transpose?
What I meant here is that it is fine for the backward to be a view even if the forward is not one.

Also if you set `TORCH_SHOW_CPP_STACKTRACES=1` while running the test, do you get more details on where the error comes from?